File: 2021/csharp/Day01.cs
public class Day01
{
    public uint[] Input { get; }

    public Day01(string inputPath)
    {
        if (!File.Exists(inputPath))
        {
            throw new Exception("File doesn't exist");
        }

        Input = File.ReadLines(inputPath).Select(line => UInt32.Parse(line)).ToArray();
    }

    public uint Part1()
    {
        uint descent = 0;
        uint previousDepth = Input[0];

        uint depth;
        for (var i = 1; i < Input.Length; i++)
        {
            if ((depth = Input[i]) > previousDepth) descent++;
            previousDepth = depth;
        }

        return descent;
    }

    public uint Part2()
    {
        uint descent = 0;
        uint[] previousDepths = { Input[0], Input[1], Input[2] };

        uint depth, shared, group, prevGroup;
        for (var i = 3; i < Input.Length; i++)
        {
            depth = Input[i];
            shared = previousDepths[1] + previousDepths[2];
            prevGroup = shared + previousDepths[0];
            group = shared + depth;
            if (group > prevGroup) descent++;
            previousDepths[0] = previousDepths[1];
            previousDepths[1] = previousDepths[2];
            previousDepths[2] = depth;
        }

        return descent;
    }
}

File: 2021/deno/day01.ts
import * as Path from "https://deno.land/std@0.116.0/path/mod.ts";
import { exec } from "./util.ts";

const scriptDir = Path.dirname(Path.fromFileUrl(import.meta.url));
const inputPath = Path.join(scriptDir, "../inputs/01.txt");

function part1(input: number[]): number {
  let descent = 0;
  let previousDepth = input[0];

  for (const depth of input.slice(1)) {
    if (depth > previousDepth) descent++;
    previousDepth = depth;
  }

  return descent;
}

function part2(input: number[]): number {
  let descent = 0;
  const previousDepths: number[] = [];

  for (const [index, depth] of input.slice(0, 3).entries()) {
    previousDepths[index] = depth;
  }

  for (const depth of input.slice(3)) {
    const shared = previousDepths[2] + previousDepths[1];
    const prevGroup = shared + previousDepths[0];
    const group = shared + depth;
    if (group > prevGroup) descent++;
    previousDepths[0] = previousDepths[1];
    previousDepths[1] = previousDepths[2];
    previousDepths[2] = depth;
  }

  return descent;
}

exec({
  inputPath,
  parseLine: (line: string) => parseInt(line),
  part1,
  part2,
}).catch((err) => console.error(err));
